### PR TITLE
Fix date_bound function for 2000 and min arguments

### PR DESCRIPTION
**Pull request by Gorovoi**

### CODE REVIEW

File: jsonknife.c [hunk 1]
```diff
@@ -665,14 +669,21 @@ date_bound(char *date_str, long str_len,  MinMax minmax){
...
 											 ObjectIdGetDatum(InvalidOid),
 											 Int32GetDatum(-1));
+
+
+
 		if(minmax == min) {
-			return min_date;
+			/* elog(INFO, "input_min: '%s', %ld", date_in, min_date); */
+
+			memcpy(date_in, &min_date, SIZEOF_DATUM);
```
Comment (rublag) [inline jsonknife.c:685]: Suspicious, directly copying Datum (uintptr_t) to char*? (1)

File: jsonknife.c [hunk 2]
```diff
@@ -729,12 +740,16 @@ date_bound(char *date_str, long str_len,  MinMax minmax){
 						(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
 						 errmsg("timestamp out of range")));
 
-			return max_date;
+			/* elog(INFO, "input_max: '%s', %ld", date_in, max_date); */
+			Datum result_max = TimestampGetDatum(max_date);
+			memcpy(date_in, &result_max, SIZEOF_DATUM);
```
Comment (rublag) [inline jsonknife.c:745]: Same. (1)

File: jsonknife.c [hunk 3]
```diff
@@ -745,28 +760,38 @@ void reduce_timestamptz(void *acc, JsonbValue *val){
...
 
 		if(dacc->minmax == min) {
-			if(dacc->acc != 0){
-				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, date, dacc->acc);
+			if(dacc->acc != NULL){
+				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, *datum, *dacc->acc);
 				/* elog(INFO, "compare %d", gt); */
 				if(gt < 0) {
-					dacc->acc = date;
+					pfree(dacc->acc);
```
Comment (rublag) [inline jsonknife.c]: pfree is not neccessary. Postgres deallocs any memory allocated with palloc when transaction ends. (2)

File: jsonknife.c [hunk 4]
```diff
@@ -745,28 +760,38 @@ void reduce_timestamptz(void *acc, JsonbValue *val){
...
 				if(gt > 0) {
-					dacc->acc = date;
+					pfree(dacc->acc);
+					dacc->acc = datum;
+				} else {
+					pfree(datum);
 				}
-			} else if (date != 0){
-				dacc->acc = date;
+			} else {
```
Comment (rublag) [inline jsonknife.c:786]: date != 0 is no longer neccessary? (3)

File: jsonknife.c [hunk 5]
```diff
@@ -785,14 +810,20 @@ knife_extract_max_timestamptz(PG_FUNCTION_ARGS) {
...
 	reduce_paths(value, paths, &acc, reduce_timestamptz);
 
-	if (acc.acc != 0)
-		PG_RETURN_NUMERIC(acc.acc);
-	else
+	if (acc.acc == NULL) {
 		PG_RETURN_NULL();
+	} else {
+		Datum result = *acc.acc;
+		pfree(acc.acc);
```
Comment (rublag) [inline jsonknife.c]: Suspicious: memory use after free. (4)

File: jsonknife.c [hunk 6]
```diff
@@ -804,28 +835,40 @@ knife_extract_min_timestamptz(PG_FUNCTION_ARGS) {
...
 	if( val != NULL && jsonbv_type(val) == JVString ) {
-		tacc->acc = accumArrayResult(tacc->acc,
-                                 (Datum) date_bound(val->val.string.val, val->val.string.len, min),
-                                 false, TIMESTAMPTZOID, CurrentMemoryContext);
+		Datum* datum = date_bound(val->val.string.val, val->val.string.len, min);
+		if (datum == NULL) {
+			tacc->acc = accumArrayResult(tacc->acc, 0, true, TIMESTAMPTZOID, CurrentMemoryContext);
+		} else {
+			Datum result = *datum;
+			pfree(datum);
```
Comment (rublag) [inline jsonknife.c]: Same. (2)

File: jsonknife.c [hunk 7]
```diff
@@ -745,28 +760,38 @@ void reduce_timestamptz(void *acc, JsonbValue *val){
...
-				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, date, dacc->acc);
+			if(dacc->acc != NULL){
+				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, *datum, *dacc->acc);
 				/* elog(INFO, "compare %d", gt); */
 				if(gt < 0) {
-					dacc->acc = date;
+					pfree(dacc->acc);
+					dacc->acc = datum;
+				} else {
+					pfree(datum);
```
Comment (rublag) [inline jsonknife.c]: Same. (2)

File: jsonknife.c [hunk 8]
```diff
@@ -745,28 +760,38 @@ void reduce_timestamptz(void *acc, JsonbValue *val){
...
+				dacc->acc = datum;
 			}
 		} else if (dacc->minmax == max ) {
-			if(dacc->acc != 0){
-				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, date, dacc->acc);
+			if(dacc->acc != NULL){
+				int gt = DirectFunctionCall2(timestamptz_cmp_timestamp, *datum, *dacc->acc);
 				if(gt > 0) {
-					dacc->acc = date;
+					pfree(dacc->acc);
```
Comment (rublag) [inline jsonknife.c]: Same. (2)

File: jsonknife.c [hunk 9]
```diff
@@ -745,28 +760,38 @@ void reduce_timestamptz(void *acc, JsonbValue *val){
...
+					dacc->acc = datum;
+				} else {
+					pfree(datum);
```
Comment (rublag) [inline jsonknife.c]: Same. (2)

File: jsonknife.c [hunk 10]
```diff
@@ -804,28 +835,40 @@ knife_extract_min_timestamptz(PG_FUNCTION_ARGS) {
...
-	else
+	if (acc.acc == NULL) {
 		PG_RETURN_NULL();
+	} else {
+		Datum result = *acc.acc;
+		pfree(acc.acc);
+
+		/* elog(INFO, "knife_extract_min_timestamptz %ld", result); */
+
+		PG_RETURN_NUMERIC(result);
```
Comment (rublag) [inline jsonknife.c]: Same. (4)

File: jsonknife.c [hunk 11]
```diff
@@ -856,11 +901,12 @@ knife_date_bound(PG_FUNCTION_ARGS) {
...
-		return res;
-	} else {
+	Datum* datum = date_bound(date, strlen(date), minmax);
+	if (datum == NULL) {
 		PG_RETURN_NULL();
 	}
+
+	Datum result = *datum;
+	pfree(datum);
+	PG_RETURN_TIMESTAMP(result);
```
Comment (rublag) [inline jsonknife.c]: Same. (4)

**Review by rublag — COMMENTED** [review]
.

File: jsonknife.c (hunk 5 above)
Comment (grafvetkindorff) [reply jsonknife.c]: move acc.acc to result (through *) and free memory. Seems ok.

File: jsonknife.c (hunk 4 above)
Comment (Gorovoi) [reply jsonknife.c:786]: 0 is '2000-01-01' date, we can't ignore it

File: jsonknife.c (hunk 1 above)
Comment (grafvetkindorff) [reply jsonknife.c:685]: copy only SIZEOF_DATUM bytes so it's ok

Comment (Gorovoi) [reply jsonknife.c:685]: Type of data doesn't matter. It copies bytes from source(`min_date`) to destination(`date_in`), both have type `void*`. We should tell compiler how it must resolve bytes in `date_in`  by casting `date_in`(`void*`) to `Datum*` after `memcpy`.  (1-r)

File: jsonknife.c (hunk 2 above)
Comment (Gorovoi) [reply jsonknife.c:745]: Same. (1-r)

File: jsonknife.c (hunk 1 above)
Comment (rublag) [reply jsonknife.c:685]: Reuse of memory is always dangerous. Performance penalty of 
Datum *datum_ptr = palloc(sizeof(*datum_ptr))
is negligible.
At least there should be MAX(..., sizeof(Datum)) in allocation to ensure that buffer overrun never happens.

Btw Datum is already a pointer. Why pointer to Datum is needed?

**Pull request closed without merging**